### PR TITLE
HDF5 specification

### PR DESCRIPTION
**Pull request by lang-m**
This notebooks describes and demonstrates the functionality of the extended HDF5 file format used in `discretisedfield`. The specification of Fields is intended to be general enough so that an extension of this document can also define a format for an HDF5 file that contains more data, e.g. a whole `drive` inside a single file, while re-using the specification (and implementation) of `discretisedfield`.

### CODE REVIEW

**Comment by lang-m** [comment]
Tests will presumably fail until ubermag/discretisedfield#377 is merged.

File: docs/hdf5-file-specification.ipynb [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+    "\n",
+    "**attributes**\n",
+    "\n",
+    "- `bc` (`str`): information about boundary conditions as defined in `discretisedfield.Mesh.bc`\n",
+    "- `n` (`list[int], length region.ndim`):  number of cells along the different spatial directions\n",
+    "\n",
+    "**datasets** (optional, only present if subregions are defined)\n",
+    "\n",
+    "- `subregion_names` (`list[str]`): names of the individual subregions defined for the field,  only present if subregions are defined\n",
+    "- `subregions` (`array[<number of subregions>, region.ndim*2]`): coordinates of `discretisedfield.Region.pmin` and `discretisedfield.Region.pmax`, subregions and subregion names are combined based on their order, only present if subregions are defined |\n",
```
Comment (samjrholt) [inline docs/hdf5-file-specification.ipynb:152]: Remove the `|`

File: docs/hdf5-file-specification.ipynb [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "93791e84-c69a-424b-9110-aa4f68587fc1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np"
```
Comment (samjrholt) [inline docs/hdf5-file-specification.ipynb:223]: Why do we import numpy again here?

File: docs/hdf5-file-specification.ipynb [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "019b5e83-c734-4466-ba58-0744f1e1518c",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# examples"
```
Comment (samjrholt) [inline docs/hdf5-file-specification.ipynb:652]: Probably remove for now

**Review by samjrholt — COMMENTED** [review]
Looks very good and informative, only a few things. One possible thing about the general structure was should talking about the attributes maybe be moved to after the tree? Then we explain basically everything in a single markdown cell rather than having a short break. Just an idea - I don't really mind :)

File: docs/hdf5-file-specification.ipynb (hunk 2 above)
Comment (lang-m) [reply docs/hdf5-file-specification.ipynb:223]: Seems to be a left-over from some checks. Removed now.

File: docs/hdf5-file-specification.ipynb [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+   "source": [
+    "# HDF5 specification in Ubermag\n",
+    "\n",
+    "## Version 0.1\n",
+    "\n",
+    "This document is valid for HDF5 files written with `ubermag >= ...` (date) TODO.\n",
+    "\n",
+    "**attributes**\n",
+    "\n",
+    "Each HDF5 file written with Ubermag has three standard attributes in the top-level group (TODO is this the correct term?).\n",
```
[the remark this answers is not in the document]
Comment (lang-m) [reply docs/hdf5-file-specification.ipynb:16]: I had a look at the h5py documentation; the correct term is `root group`; changed

File: docs/hdf5-file-specification.ipynb [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "import discretisedfield as df\n",
+    "import h5glance"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
```
[the remark this answers is not in the document]
Comment (lang-m) [reply docs/hdf5-file-specification.ipynb:41]: Removed; we do not use numpy at all.

File: docs/hdf5-file-specification.ipynb [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+   "cell_type": "markdown",
+   "id": "e2c8ac23-40d5-49e4-958d-6f5ff484ff71",
+   "metadata": {},
+   "source": [
+    "### Single field: `discretisedfield`\n",
+    "\n",
+    "An HDF5 file with `type: discretisedfield.Field` contains a single field.\n",
+    "It can be written with `field.to_file('<name>.hdf5|.h5')`.\n",
+    "\n",
+    "**additional attributes**\n",
```
[the remark this answers is not in the document]
Comment (lang-m) [reply docs/hdf5-file-specification.ipynb:61]: Changed

File: docs/hdf5-file-specification.ipynb [hunk 7]
```diff
@@ -0,0 +1,698 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "dd717fcf-ebca-48f1-ba1c-7cfb3f064fdc",
+   "metadata": {},
+   "source": [
+    "# HDF5 specification in Ubermag\n",
+    "\n",
+    "## Version 0.1\n",
+    "\n",
+    "This document is valid for HDF5 files written with `ubermag >= ...` (date) TODO.\n",
```
[the remark this answers is not in the document]
Comment (lang-m) [reply docs/hdf5-file-specification.ipynb:12]: tracked in #157

**Pull request merged**